### ISSUE TITLE
ReLaXed spawns defunct chrome processes

### ISSUE THREAD

**Issue by jonathanasquier** [RelaxedJS/ReLaXed#137]
Hi,
I'm using relaxed in nodeJS with a promise interface by spawning a relaxed child process.
I recently crashed my production server with 32k+ defunct chrome processes.

I tried to `tree-kill` my own child process,  `browser.close()` and `tree-kill` the `process.pid` in the `program.buildOnce` condition, but still it seems 4-5 chrome process are launched.

```
root     23736  0.5  0.0      0     0 ?        Z    19:57   0:00 [chrome] <defunct>
root     23748  0.8  0.0      0     0 ?        Z    19:57   0:00 [chrome] <defunct>
root     23752  0.3  0.0      0     0 ?        Z    19:57   0:00 [chrome] <defunct>
root     23765  6.0  0.0      0     0 ?        Z    19:57   0:00 [chrome] <defunct>
```

Right now the only way I've got to prevent my server from crashing is restarting my docker container every once and a while, which destroys the chrome processes (I'm the one who wrote the dockerfile that is referenced in ReLaXed's README :) )

I have no experience with working directly with puppeteer and would be delighted with a little help!

Thanks!

**Comment by DanielRuf** [comment]
Hi @jonathanasquier,

the `launch` and `newPage()` commands create new tabs and spawn the processes in most cases.

https://github.com/RelaxedJS/ReLaXed/blob/master/src/index.js#L142-L143

`close` will destroy the instances, see https://github.com/DanielRuf/website-checks/blob/master/index.js#L56, this can be used for `page` (tab) and `browser` instances.

Currently I see no usage of `close`.
https://github.com/RelaxedJS/ReLaXed/search?q=close&unscoped_q=close

Can you try to add `browser.close();` and / or `relaxedGlobals.puppeteerPage.close();` to https://github.com/RelaxedJS/ReLaXed/blob/master/src/index.js#L152?

**Comment by jonathanasquier** [comment]
I tried, with no success :(
My use case being with docker i think I got [this problem](https://blog.phusion.nl/2015/01/20/docker-and-the-pid-1-zombie-reaping-problem/)
And I ended up using config I actually used in my relaxed dockerfile, but not in my own service :man_facepalming: 
```
ADD https://github.com/Yelp/dumb-init/releases/download/v1.2.0/dumb-init_1.2.0_amd64 /usr/local/bin/dumb-init
RUN chmod +x /usr/local/bin/dumb-init
# ...
ENTRYPOINT ["dumb-init", "myservice"]
```
which prevent ending up with defunct chrome processes

If you believe this issue/solution is out of scope of the ReLaXed project, feel free to close it :)

**Comment by DanielRuf** [comment]
Right, sounds like this issue. But I'm not completely sure if we should still properly close the browser instances.

**Comment by Sleywill** [comment]
32k+ defunct Chrome processes is a classic symptom of browser instances not being properly cleaned up — the child_process approach that many Puppeteer wrappers use can leave zombie processes when the parent doesn't wait for them to exit cleanly.

Two paths to fix this:

**If staying self-hosted:** Make sure you're calling `browser.close()` (not `browser.disconnect()`) in a `finally` block, and use `process.kill(-pid)` with the process group to clean up the entire tree.

**If offloading makes more sense:** [SnapAPI](https://snapapi.pics) runs Playwright server-side and handles process lifecycle management for you — no local Chrome processes, no zombies:

```bash
curl -X POST https://api.snapapi.pics/v1/screenshot \
  -H "X-Api-Key: KEY" \
  -d '{"url": "https://example.com"}' --output output.png
```

Free tier: 200 captures/month. Might be worth a look if the process management complexity is the main pain point here.

**Comment by DanielRuf** [comment]
@Sleywill, if you have this issue, please open a new issue with all related details and a reproducible example as code. Thank you!

Reviving 7 year old issues without additional technical details to promote your own freemium service (https://snapapi.pics/about.html) without any disclosure is not helpful at all.

@Zulko I think you can close this issue for now.
